src/tree_gui.cpp

```diff
@@ -47,28 +49,61 @@ const PalSpriteID tree_sprites[] = {
  */
 class BuildTreesWindow : public Window
 {
+	/** Visual Y offset of tree root from the bottom of the tree type buttons */
+	static const int BUTTON_BOTTOM_OFFSET = 7;
+
 	uint16 base;        ///< Base tree number used for drawing the window.
 	uint16 count;       ///< Number of different trees available.
-	TreeType tree_to_plant; ///< Tree number to plant, \c TREE_INVALID for a random tree.
+	int tree_to_plant;  ///< Tree number to plant, \c TREE_INVALID for a random tree.
+
+	/**
+	 * Handle click on a tree build button.
+	 * Update GUI elements and set state to begin/end building trees.
+	 * @param tree_type Index of the button that was clicked, or -1 to abort building.
+	 */
+	void SetButtons(int tree_type)
+	{
+		/* Clicking the same type = abort placement */
+		if (this->tree_to_plant == tree_type) tree_type = -1;
+
+		if (tree_type >= 0) {
+			/* Activate placement */
+			if (_settings_client.sound.confirm) SndPlayFx(SND_15_BEEP);
+			SetObjectToPlace(SPR_CURSOR_TREE, PAL_NONE, HT_RECT, this->window_class, this->window_number);
+			this->tree_to_plant = tree_type;
+		} else {
+			/* Deactivate placement */
+			this->tree_to_plant = -1;
+			ResetObjectToPlace();
+		}
+
+		this->SetWidgetLoweredState(WID_BT_TYPE_RANDOM, this->tree_to_plant == TREE_INVALID);
+		this->GetWidget<NWidgetMatrix>(WID_BT_TYPE_MATRIX)->SetClicked(this->tree_to_plant);
+		this->SetDirty();
+	}
 
 public:
 	BuildTreesWindow(WindowDesc *desc, WindowNumber window_number) : Window(desc)
 	{
 		this->InitNested(window_number);
+		auto *matrix = this->GetWidget<NWidgetMatrix>(WID_BT_TYPE_MATRIX);
+		matrix->SetCount(this->count);
+		matrix->SetupSmallestSize(this, false);
+		this->nested_root->AssignSizePosition(ST_RESIZE, 0, 0, this->nested_root->smallest_x, this->nested_root->smallest_y, _current_text_dir == TD_RTL);
```

Comment:
I probably need to comment this bit better, about the hacks done to force the window system to calculate sizing for widgets again after calculating a minimum size for the window.